sdk/src/wallet/migration/migrate_3.rs

```diff
@@ -0,0 +1,1239 @@
+        };
+
+        if let Some(account_indexes) = storage.get::<Vec<u32>>(ACCOUNTS_INDEXATION_KEY).await? {
+            for account_index in account_indexes {
+                if let Some(mut account) = storage
+                    .get::<serde_json::Value>(&format!("{ACCOUNT_INDEXATION_KEY}{account_index}"))
+                    .await?
+                {
+                    migrate_account(&mut account)?;
+
+                    storage
+                        .set(&format!("{ACCOUNT_INDEXATION_KEY}{account_index}"), &account)
+                        .await?;
+                }
+            }
+        }
+
+        if let Some(mut wallet) = storage.get::<serde_json::Value>(WALLET_INDEXATION_KEY).await? {
+            let params = wallet["client_options"]["protocolParameters"].as_object_mut().unwrap();
+            if let Some(version) = params.remove("protocol_version") {
+                params.insert("version".to_owned(), version);
+            }
+            ConvertNetworkName::check(&mut wallet["client_options"]["protocolParameters"]["network_name"])?;
+            ConvertTokenSupply::check(&mut wallet["client_options"]["protocolParameters"]["token_supply"])?;
+            rename_keys(&mut wallet);
+
+            storage.set(WALLET_INDEXATION_KEY, &wallet).await?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+#[cfg(feature = "stronghold")]
+impl Migration<crate::client::stronghold::StrongholdAdapter> for Migrate {
+    async fn migrate(storage: &crate::client::stronghold::StrongholdAdapter) -> Result<()> {
+        use crate::{
+            client::storage::StorageAdapter,
+            wallet::core::operations::stronghold_backup::stronghold_snapshot::{ACCOUNTS_KEY, CLIENT_OPTIONS_KEY},
+        };
+
+        if let Some(mut accounts) = storage.get::<Vec<serde_json::Value>>(ACCOUNTS_KEY).await? {
+            for account in &mut accounts {
+                migrate_account(account)?;
+            }
+            storage.set(ACCOUNTS_KEY, &accounts).await?;
+        }
+
+        if let Some(mut client_options) = storage.get::<serde_json::Value>(CLIENT_OPTIONS_KEY).await? {
+            let params = client_options["protocolParameters"].as_object_mut().unwrap();
+            if let Some(version) = params.remove("protocol_version") {
+                params.insert("version".to_owned(), version);
+            }
+            ConvertNetworkName::check(&mut client_options["protocolParameters"]["network_name"])?;
+            ConvertTokenSupply::check(&mut client_options["protocolParameters"]["token_supply"])?;
+            rename_keys(&mut client_options);
+
+            storage.set(CLIENT_OPTIONS_KEY, &client_options).await?;
+        }
+        Ok(())
+    }
+}
+
+fn migrate_account(account: &mut serde_json::Value) -> Result<()> {
+    for output_data in account["outputs"]
+        .as_object_mut()
+        .ok_or(Error::Storage("malformatted outputs".to_owned()))?
```

Comment:
Can this error not just take a &str? Do we ever actually need a String?

Reply:
I dunno, but I don't want to change that in this PR

Reply:
Yeah yeah no worries, just curious